pkg/k8s/resource/diffstore.go

```diff
@@ -0,0 +1,270 @@
+
+	// Observe returns a channel which will be signalled every time there is at least one pending change.
+	// Multiple changes are coalesced. Calling `done` will drain and close the channel, it should be called when no
+	// longer using `onDiff` to free resources.
+	Observe() (onDiff <-chan struct{}, done func())
+}
+
+var _ DiffStoreFactory[*k8sRuntime.Unknown] = (*diffStoreFactory[*k8sRuntime.Unknown])(nil)
+
+type diffStoreParams[T k8sRuntime.Object] struct {
+	cell.In
+
+	LS         hive.Lifecycle
+	Shutdowner hive.Shutdowner
+	Resource   Resource[T]
+}
+
+type diffStoreFactory[T k8sRuntime.Object] struct {
+	params diffStoreParams[T]
+
+	store Store[T]
+
+	ctx      context.Context
+	cancel   context.CancelFunc
+	doneChan chan struct{}
+
+	inSync bool
+
+	mu         lock.Mutex
+	onDiff     []chan struct{}
+	diffStores []*diffStore[T]
+}
+
+// NewDiffStoreFactory
+func NewDiffStoreFactory[T k8sRuntime.Object](params diffStoreParams[T]) DiffStoreFactory[T] {
+	dsf := &diffStoreFactory[T]{
+		params:   params,
+		doneChan: make(chan struct{}),
+	}
+
+	dsf.ctx, dsf.cancel = context.WithCancel(context.Background())
+
+	params.LS.Append(dsf)
+
+	return dsf
+}
+
+// NewStore creates a new store from this factory, all DiffStores share the same underlying store but each
+// records its own diff list.
+func (dsf *diffStoreFactory[T]) NewStore() DiffStore[T] {
+	dsf.mu.Lock()
+	defer dsf.mu.Unlock()
+
+	store := &diffStore[T]{
+		factory:     dsf,
+		store:       dsf.store,
+		changedKeys: make(map[Key]bool),
+	}
+	dsf.diffStores = append(dsf.diffStores, store)
+	return store
+}
+
+// Observe returns a channel which will be signalled every time there is at least one pending change.
+// Multiple changes are coalesced. Calling `done` will drain and close the channel, it should be called when no
+// longer using `onDiff` to free resources.
+func (dsf *diffStoreFactory[T]) Observe() (onDiff <-chan struct{}, done func()) {
+	newOnDiff := make(chan struct{}, 1)
+	dsf.onDiff = append(dsf.onDiff, newOnDiff)
+	done = func() {
+		// Close
+		close(newOnDiff)
+		// And drain
+		select {
+		case <-newOnDiff:
+		default:
+		}
+	}
+	return newOnDiff, done
+}
+
+// Start implements hive.Hook
+func (dsf *diffStoreFactory[T]) Start(_ hive.HookContext) error {
+	go dsf.run()
+	return nil
+}
+
+// Stop implements hive.Hook
+func (dsf *diffStoreFactory[T]) Stop(stopCtx hive.HookContext) error {
+	dsf.cancel()
+
+	select {
+	case <-dsf.doneChan:
+	case <-stopCtx.Done():
+	}
+
+	return nil
+}
+
+func (dsf *diffStoreFactory[T]) run() {
+	defer close(dsf.doneChan)
+
+	var err error
+	dsf.store, err = dsf.params.Resource.Store(dsf.ctx)
+	if err != nil {
+		dsf.params.Shutdowner.Shutdown(hive.ShutdownWithError(err))
+		return
+	}
+
+	dsf.mu.Lock()
+	for _, store := range dsf.diffStores {
+		store.store = dsf.store
+	}
+	dsf.mu.Unlock()
+
+	errChan := make(chan error, 2)
+	updateChan := stream.ToChannel[Event[T]](dsf.ctx, errChan, dsf.params.Resource)
+
+	for {
```

Comment:
This has a bug. We might end up reading `nil` from `errChan` before consuming the last element in the stream and shutting down even when there really wasn't any error.

You should instead write this as:
```
for ev := range updateChan {
  // handle event
}
if err := <-errChan; err != nil {
  // handle stream completion error, if there was one
}
close(errChan)
```